substrate/serai/primitives/src/lib.rs

```diff
@@ -1,7 +1,31 @@
+#![cfg_attr(docsrs, feature(doc_cfg))]
+#![cfg_attr(docsrs, feature(doc_auto_cfg))]
 #![cfg_attr(not(feature = "std"), no_std)]
 
+use scale::{Encode, Decode, MaxEncodedLen};
+use scale_info::TypeInfo;
+#[cfg(feature = "std")]
+use serde::{Serialize, Deserialize};
+
+use sp_core::sr25519::{Public, Signature as RistrettoSignature};
+
 mod amount;
 pub use amount::*;
 
 mod coins;
 pub use coins::*;
+
+pub type PublicKey = Public;
+pub type NativeAddress = PublicKey;
+pub type Signature = RistrettoSignature;
+
+/// The type used to identify block numbers.
+// Doesn't re-export TendermintMachine due to traits.
+#[derive(Clone, Copy, PartialEq, Eq, Default, Debug, Encode, Decode, MaxEncodedLen, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub struct BlockNumber(pub u32);
```

Comment:
I really feel like we should be mapping whole blocks where applicable, not just relying on the block number.

Reply:
in-instructions can't use block hashes since it can't verify they're in a series. We can relax that check since its own view should be sorted? I'll look at that.

Reply:
The issue is we don't just verify the block number to be sequential. We verify a validator isn't saying block 9,999,999,999,999 just occured by ensuring the claimed block number <= our perceived block number. Replicating that check with block hashes would require making *every single hash* available under inherent data.

That's... fine? It's only actually the hashes since the last update? Though I really don't like it as we're dumping a Vec in we have to do an O(n) search on. I'd rather have In Instruction remain BlockNumber, yet use block hash for batch ID.

Reply:
Some blockchains reset their block number.